### DISCUSSION TITLE
Packing pixi environments for the Raspberry Pi

### DISCUSSION THREAD

**Discussion by stephane-caron** [upkie/upkie#467] [Tips and tricks]
If you have [configured your Upkie as a Wi-Fi access point](https://github.com/upkie/upkie/wiki/Mods#wi-fi-access-point), chances are, you have already faced the issue of installing conda packages without an Internet connection. One solution is to [connect to the Internet via a separate Wi-Fi dongle](https://github.com/orgs/upkie/discussions/114#discussioncomment-6761094), but it will become tedious if the need arises too often. In what follows, we propose a way to use the Internet connection of your host computer instead.

# Using pixi and pixi-pack

We use [pixi](https://pixi.sh/) and a tool called `pixi-pack` to export pixi environments to Upkie. If you don't have it yet, you can install pixi from [here](https://pixi.sh/latest/#installation).

First, we will want to create an `environment.tar` file. Define a `pack-to-upkie` task as done for instance in the [`pyproject.toml` of the PPO balancer](https://github.com/upkie/ppo_balancer/blob/460c1be73e0756fe5e049e9ff857d24c748a7e79/pyproject.toml), then run the following command:

```bash
pixi run pack-to-upkie
```

Once the file is created, upload it to your Upkie and unpack it by:

```bash
pixi-pack unpack environment.tar
```

If `pixi-pack` is not installed on your Upkie, you can get a `pixi-pack-aarch64-unknown-linux-gnu` binary from the [pixi-pack release page](https://github.com/Quantco/pixi-pack/releases). Finally, activate the environment and run the agent:

```bash
source ./activate.sh
python ppo_balancer/run.py
```

# See also

- This note follows a previous attempt where we [packed conda environments](https://github.com/orgs/upkie/discussions/434) (NB: using `pixi-pack` is  [much better](https://github.com/upkie/ppo_balancer/pull/5#issuecomment-2479025438))

**Comment by traversaro** [comment]
> If `pixi-pack` is not installed on your Upkie, you can get a `pixi-pack-aarch64-unknown-linux-gnu` binary from the [pixi-pack release page](https://github.com/Quantco/pixi-pack/releases).

Since 0.3.0, `pixi-pack` also support creating self-extracting archives, see https://github.com/Quantco/pixi-pack/tree/v0.3.0?tab=readme-ov-file#self-extracting-binaries , thanks to the great work of @prsabahrami and @pavelzw .
